addons/mrp/models/mrp_production.py

```diff
@@ -333,26 +333,33 @@ def get_stock_moves(moves, state):
                 ('raw_material_production_id.state', 'in', ('cancel', 'done', 'draft'))])
             return [('move_raw_ids', 'in', raw_stock_moves.ids)]
 
-        raw_stock_moves = self.env['stock.move'].search([
-            ('raw_material_production_id', '!=', False),
-            ('raw_material_production_id.state', 'not in', ('cancel', 'done', 'draft'))])
-        if operator == '=':
-            raw_stock_moves = get_stock_moves(raw_stock_moves, value)
-        elif operator == '!=':
-            raw_stock_moves = raw_stock_moves - get_stock_moves(raw_stock_moves, value)
-        elif operator == 'in':
-            search_raw_moves = self.env['stock.move']
-            for state in value:
-                search_raw_moves |= get_stock_moves(raw_stock_moves, state)
-            raw_stock_moves = search_raw_moves
-        elif operator == 'not in':
-            search_raw_moves = self.env['stock.move']
-            for state in value:
-                search_raw_moves |= raw_stock_moves - get_stock_moves(raw_stock_moves, state)
-            raw_stock_moves = search_raw_moves
-        else:
-            raise UserError(_('Operation not supported'))
-        return [('move_raw_ids', 'in', raw_stock_moves.ids)]
+        raw_productions = self.env['mrp.production'].search([('state', 'not in', ('done', 'cancel', 'draft'))])
+        selected_productions = self.env['mrp.production']
+        for prod in raw_productions:
+            raw_stock_moves = prod.move_raw_ids
+            if len(raw_stock_moves) == 0:
+                continue
+            is_production_selected = False
+            if operator == '=':
+                is_production_selected = len(get_stock_moves(raw_stock_moves, value)) == len(raw_stock_moves)
+            elif operator == '!=':
+                is_production_selected = len(get_stock_moves(raw_stock_moves, value)) == 0
+            elif operator == 'in':
+                search_raw_moves = self.env['stock.move']
+                for state in value:
+                    search_raw_moves |= get_stock_moves(raw_stock_moves, state)
+                is_production_selected = len(search_raw_moves) == len(raw_stock_moves)
+            elif operator == 'not in':
+                search_raw_moves = self.env['stock.move']
+                for state in value:
+                    search_raw_moves |= get_stock_moves(raw_stock_moves, state)
+                is_production_selected = len(search_raw_moves) == 0
+            else:
+                raise UserError(_('Operation not supported'))
+            if not is_production_selected:
+                continue
+            selected_productions += prod
```

Comment:
Another suggestion to shorten the code, but maybe it's too much compact. Keep the version you prefer
```suggestion
            raw_moves = prod.move_raw_ids
            if not raw_moves:
                continue
            if (operator == '=' and len(get_stock_moves(raw_moves, value)) == len(raw_moves)) or\
               (operator == '!=' and len(get_stock_moves(raw_moves, value)) == 0):
                selected_productions += prod
            elif operator in ['in', 'not in']:
                expected_len = len(raw_moves) if operator == 'in' else 0
                search_raw_moves = self.env['stock.move']
                for state in value:
                    search_raw_moves |= get_stock_moves(raw_moves, state)
                if len(search_raw_moves) == expected_len:
                    selected_productions += prod
```